geneweb/geneweb
# Rpc support

This PR introduces an RPC (Remote Procedure Call) client/server for Geneweb. The purpose of this feature is to allow bidirectional interaction between the database and the user without requiring page reloads.

## Organization of the PR
The feature is structured as follows:
-  Two executables `bin/rpc-client` and `bin/rpc-server` and a common library `lib/rpc`.
-  A new package `geneweb-rpc` that contains the library and the two executables. As this feature introduces several new dependencies, user can reduce the size of their opam switch by installing only `geneweb` if they don't want the full-text search feature.
- The `rpc-client` is Javascript client written in Js_of_ocaml. A short example is given in `test/rpc.html`. The client establishes a persistent WebSocket connection to the server as follows:
```javascript
const client = new Rpc("ws://localhost:8080/pingpong");
```
Here, `client` is a promise, and remote procedures can be invoked as if they were local functions. In this example, the server provides a remote produced named `ping`. To invoke it, type `client.ping()`. These calls are asynchronous and non-blocking (even though it is not recommended to send slow requests to the server for design reasons).
- The `rpc-server` is a WebSocket server with support for TLS. You can also limit the number of simultaneous connection and automatically closed idle connections. The server exchanges using JSON-RPC messages.

## How to test this PR
To test the server, start it with the following command (you need to install extra dependencies before):
```
dune exec -- _build/default/bin/rpc-server/main.exe --debug rpc
```
This will start the server (without TLS by default) on `localhost:8080`. Note that the debug flag should not be used in production environments.

You can ping the server with the test client:
```
dune exec -- _build/default/test/rpc_test.exe
```
Or load the HTML page `test/rpc.html` to test the Javascript client.
